### ISSUE TITLE
[META] Should we stop using the farmOS GitHub issue queue?

### ISSUE THREAD

**Issue by mstenta** [farmOS/farmOS#374]
*I'm creating this issue in the GitHub issue queue in hopes that it gets the attention of anyone watching the farmOS GitHub repository. :-)*

**TL;DR: The farmOS GitHub issue queue has always been a bit out-of-place in the larger farmOS project management scheme. Should we shut it down and direct people to the forum instead?**

farmOS began on Drupal.org (https://drupal.org/project/farm), and we still use [that issue queue](https://drupal.org/project/issues/farm) as the canonical place for farmOS **core** roadmaps and development tasks. Drupal.org offers issue management features beyond what GitHub provides, in a way that is tailored to Drupal codebases. There are no plans of moving off of Drupal.org.

*I say "**core**" because I want to be clear: I am ONLY referring to the farmOS SERVER codebase (this repository). Not [Field Kit](https://github.com/farmOS/farmOS-client), or the [Aggregator](https://github.com/farmOS/farmOS-aggregator), or [farmOS.js](https://github.com/farmOS/farmOS.js)/[farmOS.py](https://github.com/farmOS/farmOS.py)/[farmOS-map](https://github.com/farmOS/farmOS-map). Those are separate projects/repositories, and they have their own canonical issue queues already (only on GitHub).*

The farmOS GitHub repository was originally set up as a mirror, and the thought was that its issue queue could be used as a secondary place to post things, but that any big changes would be spun off as Drupal.org issues. A lot of developers have GitHub accounts, so it lowered the barrier to participation.

It has served us well, bringing a lot of GitHub users into the farmOS community (and bringing some more farmers to GitHub!) - but it has also created a fractured issue management workflow, which adds overhead to the maintainers. And it creates confusion to newcomers looking to participate. Most don't even know that https://drupal.org/project/issues/farm exists.

I always thought of the GitHub issue queue as somewhat of a "forum" - a place for general participation, and small features/bug reports. But then we created an actual forum: https://farmOS.discourse.group

Should we close all issues and put up some kind of message to direct people to the forum instead? And when forum discussions warrant new issues, we create those in the relevant repository-specific issue queues, whether that be on drupal.org (for farmOS core), or on GitHub (for our other projects)? Would this help to centralize discussions and encourage everyone start conversations in the same place? What are the disadvantages of this approach? Other thoughts/considerations? I want to hear what some of the frequent farmOS issue queue users think. :-)

**Comment by mstenta** [comment]
Apart from the issue queue, I would still like to allow pull requests. It is a useful way to contribute and collaborate on branches.

**Comment by symbioquine** [comment]
At least from my perspective, the Drupal.org side of things is the odd-man-out.

Working with Github issues and Markdown are general purpose skills that apply to ***many*** open source projects, whereas Drupal.org is likely to only be used by the subset of projects that happen to be connected with Drupal. This means it will always be lower leverage for users/developers to familiarize themselves with the Drupal.org issue management system and its markup than Github (or other general-purpose collaboration platforms).

Beyond my preference for Markdown, Github also does a few other things better as far as I can tell;

* I can use my clipboard to paste images and have them automatically uploaded and inserted with the correct markup
* I get inline status information about pull requests which are linked by - or reference - an issue
* Syntax highlighting support for languages other than PHP

**Comment by mstenta** [comment]
I agree with all of these points 100% @symbioquine 

That said, the question here is not which system is *better*, but rather whether or not we should have TWO systems running in parallel.

The Drupal.org issue queue is the current canonical queue for farmOS core. Changing that would be a much bigger undertaking, and one that I would argue against. If we want, we could start another discussion to work through and document that reasoning. And maybe that discussion is a prerequisite of this one.

And worth reiterating, to be clear:

> _I am ONLY referring to the farmOS SERVER codebase (this repository). Not Field Kit, or the Aggregator, or farmOS.js/farmOS.py/farmOS-map. Those are separate projects/repositories, and they have their own canonical issue queues already (only on GitHub)._

**Comment by mstenta** [comment]
Put another way: do the benefits of having a second issue queue on GitHub outweigh the costs of managing two issue queues in parallel?

Maybe they do! This is something I've had on my mind for years, so I felt it was time to make it an official question to the rest of the community.

Don't worry - I'm not planning on making any major changes like this on short notice. :-)

**Comment by symbioquine** [comment]
> That said, the question here is not which system is _better_, but rather whether or not we should have TWO systems running in parallel.

I wasn't so much trying to determine which is better in general, but describe why it is more compelling for me (and I think many other users/developers) to engage on Github issues than Drupal.org issues. The Discourse forum addresses some of the posting/images/Markup points, but it doesn't have the same inertia since it is a separate site/registration/notification-preference/etc.

All else aside I wouldn't advocate for having more than one parallel issue tracking system, but I haven't seen much to argue for that system being the Drupal.org one - other than it being the incumbent.

**Comment by mstenta** [comment]
Some related discussion in #farmOS chat today: https://irc.farmos.org/bot/log/farmOS/2020-11-22

**Comment by mstenta** [comment]
There is a middle ground here too.

We could keep the issue queue open, but make more explicit policies around what it is used for, and be more strict about closing issues and moving them to other places.

I think the main thing I struggle with is having two sets of "open" issues for the same codebase.

**Comment by symbioquine** [comment]
Based on the discussion in the Matrix channel and here, I think I favor a hybrid strategy where Github issues are used to accept new bugs/suggestions/etc but those are triaged pretty aggressively. It almost certainly doesn't make sense to keep long-term tracking issues or lingering discussions in two issue queues.

**Comment by mstenta** [comment]
I think this sounds right to me too. And ultimately, it just means we've landed right back where we started. :-)

This is essentially the approach I've been taking in my own head, although not very strictly or diligently.

> a hybrid strategy where Github issues are used to accept new bugs/suggestions/etc but those are triaged pretty aggressively

The "triaged aggressively" piece is the hard part here, and ultimately is why I raised this question in the first place. It requires work to maintain both queues. But it does seem like the benefits outweigh the costs.

Perhaps a good first step would be to create a GitHub issue template in this repo, with some guidance for newcomers. I'll take a quick pass at drafting something up and link it here for review...

**Comment by mstenta** [comment]
Here's an idea... what if we try to limit the https://github.com/farmOS/farmOS/issues queue to "bug reports" primarily? And point users to the forum for other things (or directly to specific project queues if they know where it should go).

The GitHub issue template feature allows creating JUST a bug report template. It's even possible to restrict so that bug report is the ONLY option (although I'm not sure we would need to be that strict).

Perhaps a single bug report issue template, combined with a `CONTRIBUTING.md` file for newcomers (which we could probably just migrate from https://github.com/farmOS/farmOS.org/blob/7.x-1.x/docs/community/contribute.md) would be enough?

I might sketch these ideas up in a PR to see what it might look like... and get your thoughts...

**Comment by mstenta** [comment]
> what if we try to limit the https://github.com/farmOS/farmOS/issues queue to "bug reports" primarily?

Finishing this thought... In our conversation in chat yesterday, @paul121 made a good point that bug reports are an example of something that is tricky for newcomers, because they probably don't know what component of the code is responsible, and posting them to the forum is weird (albeit fine, and also welcome).

So: I could see a system in which bug reports start in the GitHub issue bucket, and we triage that by moving them to other repos (or quick fix and close etc). So we still keep this queue minimal, and try to direct other discussions to the forum as a general starting place. Feature requests that are discussed in the forum can then be broken up into specific tasks and roadmaps in the appropriate issue queues for specific projects.

**Comment by mstenta** [comment]
And: specifically with the farmOS core codebase, we would still accept pull requests via GitHub AND merge requests/patches via Drupal.org. So whatever workflow contributors prefer would still be accepted. A GitHub PR can have conversations very similar to issues, so it would be fine to allow some things to happen in that side. Most core maintainer issue tracking can still be in Drupal.org.

**Comment by mstenta** [comment]
Pull requests for review: https://github.com/farmOS/farmOS/pull/376 + https://github.com/farmOS/farmOS.org/pull/102

**Comment by symbioquine** [comment]
@mstenta Left some minor feedback on #376

**Comment by mstenta** [comment]
I merged the `CONTRIBUTING.md` - but for some reason it isn't showing up as a link at the top of https://github.com/farmOS/farmOS/issues/

Maybe something else is required there...

I also used `ISSUE_TEMPLATE/config.yml` to add a link to the forum! Neat! https://github.com/farmOS/farmOS/issues/new/choose

**Comment by mstenta** [comment]
I submitted a support ticket to GitHub about the `CONTRIBUTING.md` issue - I followed their instructions verbatim here: https://docs.github.com/en/free-pro-team@latest/github/building-a-strong-community/setting-guidelines-for-repository-contributors

Also, seems like some other repos are experiencing the same issue, like OpenLayers: https://github.com/openlayers/openlayers/blob/main/CONTRIBUTING.md

**Comment by mstenta** [comment]
After speaking with @symbioquine and @paul121 I think we will close this issue. The main question has been answered: we will NOT stop using this issue queue. :-)

But, we've taken some concrete steps to encourage people to post to the forum instead, and use this mainly for bug reports. I'm happy with that.

We can work on triaging/moving the currently open issues moving forward... and close this issue.

Thanks for the input and discussion!

**Comment by mstenta** [comment]
> I submitted a support ticket to GitHub about the CONTRIBUTING.md issue - I followed their instructions verbatim here: https://docs.github.com/en/free-pro-team@latest/github/building-a-strong-community/setting-guidelines-for-repository-contributors

FYI I received a response about this:

> The link will only show up for first time contributors. Since you've already contributed to that repository you will not see the link.

I didn't know it worked like that. Makes sense I suppose. I'd rather it get displayed consistently even to old contributors. It's nice to have a consistent link in a predictable place. But I suppose this does what we need.

**Comment by symbioquine** [comment]
From another account I can see;

![image](https://user-images.githubusercontent.com/30754460/101285799-06984380-379c-11eb-8877-b60cf29275db.png)
